apps/extension/src/shared/components/grpc-endpoint-form/use-grpc-endpoint-form.ts

```diff
@@ -0,0 +1,117 @@
+import { Code, ConnectError, createPromiseClient } from '@connectrpc/connect';
+import { QueryService } from '@buf/penumbra-zone_penumbra.connectrpc_es/penumbra/core/app/v1/app_connect';
+import { createGrpcWebTransport } from '@connectrpc/connect-web';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { AllSlices } from '../../../state';
+import { useStoreShallow } from '../../../utils/use-store-shallow';
+import { ServicesMessage } from '@penumbra-zone/types/src/services';
+import { GRPC_ENDPOINTS } from '@penumbra-zone/constants/src/grpc-endpoints';
+import { debounce } from 'lodash';
+
+const randomSort = () => (Math.random() >= 0.5 ? 1 : -1);
+
+const useSaveGrpcEndpointSelector = (state: AllSlices) => ({
+  grpcEndpoint: state.network.grpcEndpoint,
+  setGrpcEndpoint: state.network.setGRPCEndpoint,
+});
+
+const isValidUrl = (url: string) => {
+  try {
+    new URL(url);
+    return true;
+  } catch {
+    return false;
+  }
+};
+
+export const useGrpcEndpointForm = () => {
```

Comment:
This hook basically extracts all the business logic out of the `SettingsRPC` component

Reply:
Yeah, you're right — I'll do a follow-up PR for that